Local plugin "Unable to compile TypeScript" error with ts-node (without swc)

### Current Behavior

While https://github.com/nrwl/nx/issues/9823 is solved (by https://github.com/nrwl/nx/pull/14995), in my workspace without swc and relying on ts-node instead, I'm getting a new error (see below _Failure Logs_)



### Expected Behavior

The executor should successfully run, as it does when swc is installed:

```
» nx run demo:build

> nx run demo:build

Executor ran for Build {}

 —————————————————————————————————————————————————————————————

 >  NX   Successfully ran target build for project demo (378ms)
```

### GitHub Repo

https://github.com/forivall/nx-repro-local-plugin-issue-without-swc

### Steps to Reproduce

1. `npm install`
2. `npx nx run demo:build --verbose`


### Nx Report

```shell
Node : 16.17.0
OS   : darwin arm64
npm  : 8.15.0

nx                      : 15.7.1
@nrwl/jest              : 15.7.1
@nrwl/linter            : 15.7.1
@nrwl/workspace         : 15.7.1
@nrwl/cli               : 15.7.1
@nrwl/devkit            : 15.7.1
@nrwl/eslint-plugin-nx  : 15.7.1
@nrwl/js                : 15.7.1
@nrwl/nx-plugin         : 15.7.1
@nrwl/tao               : 15.7.1
typescript              : 4.8.4
---------------------------------------
Local workspace plugins:
 @repro-local-plugin-issue/repro
```


### Failure Logs

```shell
» nx run demo:build --verbose

> nx run demo:build

 >  NX   ⨯ Unable to compile TypeScript:
   error TS5024: Compiler option 'module' requires a value of type string.
   error TS5024: Compiler option 'target' requires a value of type string.

TSError: ⨯ Unable to compile TypeScript:
error TS5024: Compiler option 'module' requires a value of type string.
error TS5024: Compiler option 'target' requires a value of type string.
    at createTSError (/Users/emily.klassen/code/sandbox/repro-local-plugin-issue/node_modules/ts-node/src/index.ts:859:12)
    at reportTSError (/Users/emily.klassen/code/sandbox/repro-local-plugin-issue/node_modules/ts-node/src/index.ts:863:19)
    at createFromPreloadedConfig (/Users/emily.klassen/code/sandbox/repro-local-plugin-issue/node_modules/ts-node/src/index.ts:874:36)
    at create (/Users/emily.klassen/code/sandbox/repro-local-plugin-issue/node_modules/ts-node/src/index.ts:624:10)
    at register (/Users/emily.klassen/code/sandbox/repro-local-plugin-issue/node_modules/ts-node/src/index.ts:591:15)
    at registerTranspiler (/Users/emily.klassen/code/sandbox/packages/nx/src/utils/register.ts:64:25)
    at registerTranspiler (/Users/emily.klassen/code/sandbox/packages/nx/src/utils/register.ts:78:12)
    at registerPluginTSTranspiler (/Users/emily.klassen/code/sandbox/packages/nx/src/utils/nx-plugin.ts:194:23)
    at /Users/emily.klassen/code/sandbox/packages/nx/src/config/workspaces.ts:326:35
    at /Users/emily.klassen/code/sandbox/packages/nx/src/command-line/run.ts:189:28

 ————————————————————————————————————————————————————————————————————————————

 >  NX   Running target build for project demo failed

   Failed tasks:

   - demo:build

   Hint: run the command with --verbose for more details.
```


### Additional Information

_No response_

--- a/e2e/nx-plugin/src/nx-plugin.test.ts
+++ b/e2e/nx-plugin/src/nx-plugin.test.ts
@@ -15,7 +15,11 @@ import {
   readFile,
   removeFile,
   cleanupProject,
+  runCommand,
+  getPackageManagerCommand,
+  updateJson,
 } from '@nrwl/e2e/utils';
+import type { PackageJson } from 'nx/src/utils/package-json';
 
 import { ASYNC_GENERATOR_EXECUTOR_CONTENTS } from './nx-plugin.fixtures';
 
@@ -359,6 +363,30 @@ describe('Nx Plugin', () => {
 
       expect(() => checkFilesExist(`libs/${generatedProject}`)).not.toThrow();
       expect(() => runCLI(`execute ${generatedProject}`)).not.toThrow();
+    });
+
+    it('should work with ts-node only', async () => {
+      const oldPackageJson: PackageJson = readJson('package.json');
+      updateJson<PackageJson>('packageJson', (j) => {
+        delete j.dependencies['@swc-node/register'];
+        delete j.devDependencies['@swc-node/register'];
+        return j;
+      });
+      runCommand(getPackageManagerCommand().install);
+
+      const generator = uniq('generator');
+
+      expect(() => {
+        runCLI(
+          `generate @nrwl/nx-plugin:generator ${generator} --project=${plugin}`
+        );
+
+        runCLI(
+          `generate @${npmScope}/${plugin}:${generator} --name ${uniq('test')}`
+        );
+      }).not.toThrow();
+      updateFile('package.json', JSON.stringify(oldPackageJson, null, 2));
+      runCommand(getPackageManagerCommand().install);
     });
   });
 

--- a/e2e/nx-plugin/src/nx-plugin.test.ts
+++ b/e2e/nx-plugin/src/nx-plugin.test.ts
@@ -367,7 +367,7 @@ describe('Nx Plugin', () => {
 
     it('should work with ts-node only', async () => {
       const oldPackageJson: PackageJson = readJson('package.json');
-      updateJson<PackageJson>('packageJson', (j) => {
+      updateJson<PackageJson>('package.json', (j) => {
         delete j.dependencies['@swc-node/register'];
         delete j.devDependencies['@swc-node/register'];
         return j;

--- a/packages/nx/src/utils/register.spec.ts
+++ b/packages/nx/src/utils/register.spec.ts
@@ -1,0 +1,20 @@
+import { ModuleKind, ScriptTarget } from 'typescript';
+import { getTsNodeCompilerOptions } from './register';
+
+describe('getTsNodeCompilerOptions', () => {
+  it('should replace enum value with enum key for module', () => {
+    expect(
+      getTsNodeCompilerOptions({
+        module: ModuleKind.CommonJS,
+      }).module
+    ).toEqual('CommonJS');
+  });
+
+  it('should replace enum value with enum key for target', () => {
+    expect(
+      getTsNodeCompilerOptions({
+        target: ScriptTarget.ES2020,
+      }).target
+    ).toEqual('ES2020');
+  });
+});

--- a/packages/nx/src/utils/register.ts
+++ b/packages/nx/src/utils/register.ts
@@ -1,6 +1,5 @@
 import { dirname, join } from 'path';
-import { TsConfigOptions } from 'ts-node';
-import type { CompilerOptions, TsConfigSourceFile } from 'typescript';
+import type { CompilerOptions } from 'typescript';
 import { logger, NX_PREFIX, stripIndent } from './logger';
 
 const swcNodeInstalled = packageIsInstalled('@swc-node/register');
@@ -174,31 +173,31 @@ function packageIsInstalled(m: string) {
  * `register`'s signature just types the field as `object`, so we
  * unfortunately do not get any kind of type safety on this.
  */
-function getTsNodeCompilerOptions(compilerOptions: CompilerOptions) {
+export function getTsNodeCompilerOptions(compilerOptions: CompilerOptions) {
   if (!ts) {
     ts = require('typescript');
   }
-  return {
-    ...compilerOptions,
-    ...getValueFromTsEnum(compilerOptions, 'module', 'ModuleKind'),
-    ...getValueFromTsEnum(compilerOptions, 'target', 'ScriptTarget'),
-    ...getValueFromTsEnum(
-      compilerOptions,
-      'moduleDetection',
-      'ModuleDetectionKind'
-    ),
-    ...getValueFromTsEnum(compilerOptions, 'newLine', 'NewLineKind'),
-    ...getValueFromTsEnum(
-      compilerOptions,
-      'moduleResolution',
-      'ModuleResolutionKind'
-    ),
-    ...getValueFromTsEnum(
-      compilerOptions,
-      'importsNotUsedAsValues',
-      'ImportsNotUsedAsValues'
-    ),
+
+  const flagMap: Partial<
+    Record<keyof RemoveIndex<CompilerOptions>, keyof typeof ts>
+  > = {
+    module: 'ModuleKind',
+    target: 'ScriptTarget',
+    moduleDetection: 'ModuleDetectionKind',
+    newLine: 'NewLineKind',
+    moduleResolution: 'ModuleResolutionKind',
+    importsNotUsedAsValues: 'ImportsNotUsedAsValues',
   };
+
+  const result = { ...compilerOptions };
+
+  for (const flag in flagMap) {
+    if (compilerOptions[flag]) {
+      result[flag] = ts[flagMap[flag]][compilerOptions[flag]];
+    }
+  }
+
+  return result;
 }
 
 /**
@@ -209,15 +208,3 @@ function getTsNodeCompilerOptions(compilerOptions: CompilerOptions) {
 type RemoveIndex<T> = {
   [K in keyof T as {} extends Record<K, 1> ? never : K]: T[K];
 };
-
-function getValueFromTsEnum(
-  compilerOptions: CompilerOptions,
-  flag: keyof RemoveIndex<CompilerOptions>,
-  enumName: keyof typeof ts
-) {
-  return compilerOptions[flag]
-    ? {
-        [flag]: ts[enumName][compilerOptions[flag] as any],
-      }
-    : {};
-}

--- a/packages/nx/src/utils/register.ts
+++ b/packages/nx/src/utils/register.ts
@@ -1,9 +1,11 @@
 import { dirname, join } from 'path';
-import type { CompilerOptions } from 'typescript';
+import { TsConfigOptions } from 'ts-node';
+import type { CompilerOptions, TsConfigSourceFile } from 'typescript';
 import { logger, NX_PREFIX, stripIndent } from './logger';
 
 const swcNodeInstalled = packageIsInstalled('@swc-node/register');
 const tsNodeInstalled = packageIsInstalled('ts-node/register');
+let ts: typeof import('typescript');
 
 /**
  * Optionally, if swc-node and tsconfig-paths are available in the current workspace, apply the require
@@ -63,7 +65,7 @@ export function registerTranspiler(
       registerTranspiler = () => {
         const service = register({
           transpileOnly: true,
-          compilerOptions,
+          compilerOptions: getTsNodeCompilerOptions(compilerOptions),
         });
         // Don't warn if a faster transpiler is enabled
         if (!service.options.transpiler && !service.options.swc) {
@@ -121,8 +123,10 @@ function readCompilerOptions(tsConfigPath): CompilerOptions {
 }
 
 function readCompilerOptionsWithTypescript(tsConfigPath) {
-  const { readConfigFile, parseJsonConfigFileContent, sys } =
-    require('typescript') as typeof import('typescript');
+  if (!ts) {
+    ts = require('typescript');
+  }
+  const { readConfigFile, parseJsonConfigFileContent, sys } = ts;
   const jsonContent = readConfigFile(tsConfigPath, sys.readFile);
   const { options } = parseJsonConfigFileContent(
     jsonContent,
@@ -163,4 +167,57 @@ function packageIsInstalled(m: string) {
   } catch {
     return false;
   }
+}
+
+/**
+ * ts-node requires string values for enum based typescript options.
+ * `register`'s signature just types the field as `object`, so we
+ * unfortunately do not get any kind of type safety on this.
+ */
+function getTsNodeCompilerOptions(compilerOptions: CompilerOptions) {
+  if (!ts) {
+    ts = require('typescript');
+  }
+  return {
+    ...compilerOptions,
+    ...getValueFromTsEnum(compilerOptions, 'module', 'ModuleKind'),
+    ...getValueFromTsEnum(compilerOptions, 'target', 'ScriptTarget'),
+    ...getValueFromTsEnum(
+      compilerOptions,
+      'moduleDetection',
+      'ModuleDetectionKind'
+    ),
+    ...getValueFromTsEnum(compilerOptions, 'newLine', 'NewLineKind'),
+    ...getValueFromTsEnum(
+      compilerOptions,
+      'moduleResolution',
+      'ModuleResolutionKind'
+    ),
+    ...getValueFromTsEnum(
+      compilerOptions,
+      'importsNotUsedAsValues',
+      'ImportsNotUsedAsValues'
+    ),
+  };
+}
+
+/**
+ * Index keys allow empty objects, where as "real" keys
+ * require a value. Thus, this filters out index keys
+ * See: https://stackoverflow.com/a/68261113/3662471
+ */
+type RemoveIndex<T> = {
+  [K in keyof T as {} extends Record<K, 1> ? never : K]: T[K];
+};
+
+function getValueFromTsEnum(
+  compilerOptions: CompilerOptions,
+  flag: keyof RemoveIndex<CompilerOptions>,
+  enumName: keyof typeof ts
+) {
+  return compilerOptions[flag]
+    ? {
+        [flag]: ts[enumName][compilerOptions[flag] as any],
+      }
+    : {};
 }

